core/dive.c

```diff
@@ -2232,36 +2233,57 @@ static int match_cylinder(const cylinder_t *cyl, const struct dive *dive, const
  * use, but we might want to fill in any missing cylinder details
  * in 'a' if we had it from 'b'.
  */
-static void merge_one_cylinder(struct cylinder_table *t, const cylinder_t *a, const cylinder_t *b)
-{
-	cylinder_t *res = add_empty_cylinder(t);
-	res->type.size.mliter = a->type.size.mliter ?
-		a->type.size.mliter : b->type.size.mliter;
-	res->type.workingpressure.mbar = a->type.workingpressure.mbar ?
-		a->type.workingpressure.mbar : b->type.workingpressure.mbar;
-	res->type.description = !empty_string(a->type.description) ?
-		copy_string(a->type.description) : copy_string(b->type.description);
-	res->gasmix = a->gasmix;
-	res->start.mbar = a->start.mbar ?
-		a->start.mbar : b->start.mbar;
-	res->end.mbar = a->end.mbar ?
-		a->end.mbar : b->end.mbar;
+static void merge_one_cylinder(cylinder_t *a, const cylinder_t *b)
+{
+	if (!a->type.size.mliter)
+		a->type.size.mliter = b->type.size.mliter;
+	if (!a->type.workingpressure.mbar)
+		a->type.workingpressure.mbar = b->type.workingpressure.mbar;
+	if (empty_string(a->type.description))
+		a->type.description = copy_string(b->type.description);
```

Comment:
Probably a stupid comment, however note that empty_string returns true for a pointer to "". That would result in a tiny memory leak. I'm not sure if that can actually happen.